### PR TITLE
Added WithRetryPeriodFunc option

### LINKED ISSUE #62: WithRetryPeriod overload to enable custom retry strategies
**Background**
When services loose their connection, they will retry using some constant time.duration which might lead to a thundering herd if there are enough services that retry at the same time.
It can to some extend be mitigated by introducing a random jitter, but strategies like exponential backoff could also be relevant

**Current API limitations**
In the place where I work we are using the function:
```go
func WithRetryPeriod(dur time.Duration) Option
```

which only allows us to set a jitter. We see benefits in using an exponential backoff strategy instead.

**Proposal**
Would it be possible to get an overload like the following:
```go
func WithRetryPeriodFunc(timeFunc func(retryCount int) dur time.Duration) Option
```

That would make it possible for us to choose our own retry strategy, without adding the specific strategy to amqpextra.

I will be happy to give the implementation a go

### PR DESCRIPTION
**Pull request by mahlunar**
**Added `WithRetryPeriodFunc` option to be able to change the retry stategy.**

**Changes to consumer**
- Added func Option
- Renamed `retryPeriod`field to `nextRetryPeriod` to indicate intent
- waitRetry method now calls the retryCounter to get the current attempt
- the retrycounter channel is created and registered into the `stateChs` field.

**Additions**
- Added a retryCounter struct which runs its own internal go routine that updates a counter via its channel that receives state change events from the consumer.

I did not find code that closes the `stateChs` channels so I chose to close the retryCounter channel when the loop returns.

I have added tests for the `retryCounter` where I ran into a problem regarding reads on the counter that were not updated before the tests asserted on the value. I have added an `assertRetryCount` which will run a loop that breaks after 10 ms or when the assertion succeeds. Let me know what you think.

Fixes #62

### CODE REVIEW

**Comment by mahlunar** [comment]
I can see that there is a race on the channel in the `retryCounter` introduced. What should we do here? Would it make sense to **not** close the channel when the loop returns?

**Comment by makasim** [comment]
What about a bit simpler solution? 

* Let's introduce `retryCount int` property in Consumer.
* When we are in retry loop we constantly hit `waitRetry` func. We can calc wait duration and safely increment `retryCount`. 
* in `consumerState` [here](https://github.com/makasim/amqpextra/blob/350c5620b4add78833bef5721e969c72cef0a2fa/consumer/consumer.go#L387) we can set the counter to zero. 

There wont be any races cause only one gouroutin access that property.

WDYT?

**Comment by mahlunar** [comment]
> What about a bit simpler solution?
> 
> * Let's introduce `retryCount int` property in Consumer.
> * When we are in retry loop we constantly hit `waitRetry` func. We can calc wait duration and safely increment `retryCount` there.
> * in `consumerState` [here](https://github.com/makasim/amqpextra/blob/350c5620b4add78833bef5721e969c72cef0a2fa/consumer/consumer.go#L387) we can set the counter to zero.
> 
> There wont be any races cause only one gouroutin access that property.
> 
> WDYT?

That would be great. I was also thinking about a similar solution, but was a bit in doubt where it was safe to reset the counter. I like the solution. 👌🏼  I will do the necessary changes.

File: consumer/consumer.go [hunk 1]
```diff
@@ -419,14 +430,18 @@ func (c *Consumer) consumeState(ch AMQPChannel, queue string, connCloseCh <-chan
...
+	timer := time.NewTimer(c.nextRetryPeriod(c.retryCounter))
 	defer func() {
 		timer.Stop()
 		select {
 		case <-timer.C:
 		default:
 		}
 	}()
+	defer func() {
+		c.retryCounter++
```
Comment (makasim) [inline consumer/consumer.go:442]: IMO should be ok without defer

**Comment by makasim** [comment]
thanks!

**Pull request merged**